From b9b8ed00c9ddcb54df04616f8c72b7e257169261 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 27 Jun 2022 15:23:56 +0200
Subject: [PATCH 168/196] Fix `GetBlockSubsidy` for difficulty < ~16 (#260)

Currently, this function asserts nBits <= 0x1c100000, but the on the testnet, this can now fall below that, so we simply cap the subsidy in this case.
---
 src/test/validation_tests.cpp | 2 ++
 src/validation.cpp            | 5 ++---
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/test/validation_tests.cpp b/src/test/validation_tests.cpp
index c37a9c7ae..6fee42623 100644
--- a/src/test/validation_tests.cpp
+++ b/src/test/validation_tests.cpp
@@ -85,6 +85,8 @@ static void TestBlockSubsidyLogarithmic(const Consensus::Params &params) {
         {0x1000ffff, int64_t(26'260'005'725) * SATOSHI},
         {0x0500ffff, int64_t(49'140'005'725) * SATOSHI},
         {0x0300ffff, int64_t(53'300'005'725) * SATOSHI},
+        {0x1c100001, 1 * SUBSIDY}, // for testnet
+        {0x1d00ffff, 1 * SUBSIDY}, // for testnet
     };
     for (const auto &pair : test_cases) {
         const Amount subsidy = GetBlockSubsidy(pair.first, params);
diff --git a/src/validation.cpp b/src/validation.cpp
index 4ca1dbb1c..ae0e5764a 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -862,7 +862,8 @@ static bool WriteBlockToDisk(const CBlock &block, FlatFilePos &pos,
 Amount GetBlockSubsidy(uint32_t nBits,
                        const Consensus::Params &consensusParams) {
     static constexpr int64_t SUBSIDY_INT = SUBSIDY / SATOSHI;
-    if (!consensusParams.enableDifficultyBasedSubsidy) {
+    // Always cap on regtest, and sometimes we need to cap on testnet.
+    if (!consensusParams.enableDifficultyBasedSubsidy || nBits > 0x1c100000) {
         return SUBSIDY;
     }
     // We define:
@@ -889,8 +890,6 @@ Amount GetBlockSubsidy(uint32_t nBits,
     static constexpr int64_t LOG_MAX_TARGET = 220;
     const int64_t exp = nBits >> 24;
     const uint32_t mantissa = (nBits & 0x007fffff);
-    // CheckProofOfWork ensures this
-    assert(nBits <= 0x1c100000);
     // First check in ContextualCheckBlockHeader ensures this (since nBits have
     // to match exactly, enforcing canonical encoding)
     assert(mantissa >= 0x8000);
-- 
2.39.5 (Apple Git-154)

